### PR TITLE
Take only Assembly instance as an input to ConstantOptimiser

### PR DESCRIPTION
**Pull request by axic**
I'm not sure which is the better approach. Neither the old or the new one are perfect.

### CODE REVIEW

**Review by leonardoalt — DISMISSED** [review]

File: libevmasm/Assembly.h [hunk 1]
```diff
@@ -86,6 +86,9 @@ class Assembly
 	/// Returns the assembly items.
 	AssemblyItems const& items() const { return m_items; }
 
+	/// Returns the mutable assembly items.
+	AssemblyItems& itemsMutable() { return m_items; }
```
Comment (axic) [inline libevmasm/Assembly.h]: Any alternative notation `Assembly.h` uses is to overload: `AssemblyItems& items() {..}`.

Comment (chriseth) [reply libevmasm/Assembly.h]: I actually think that a const_cast in a test is better than providing a mutable getter in the general interface.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libevmasm/Assembly.h (hunk 1 above)
Comment (axic) [reply libevmasm/Assembly.h]: That was the last option I've considered. I can change it, but note, the Assembly interface already exports immutable and mutable versions of the same in another case:
```c++
        Assembly const& sub(size_t _sub) const { return *m_subs.at(_sub); }
        Assembly& sub(size_t _sub) { return *m_subs.at(_sub); }
```

Comment (chriseth) [reply libevmasm/Assembly.h]: Yes, of the assembly interface, but not of the underlying assembly item array, or does it?

Comment (axic) [reply libevmasm/Assembly.h]: Changed it. I'm not sure I agree with the design decision, but its here to be merged.

**Comment by axic** [comment]
@chriseth updated

File: libevmasm/ConstantOptimiser.cpp [hunk 2]
```diff
@@ -30,10 +30,12 @@ unsigned ConstantOptimisationMethod::optimiseConstants(
 	bool _isCreation,
 	size_t _runs,
 	solidity::EVMVersion _evmVersion,
-	Assembly& _assembly,
-	AssemblyItems& _items
+	Assembly& _assembly
 )
 {
+	// TODO: design the optimiser in a way this is not needed
+	AssemblyItems& _items = const_cast<AssemblyItems &>(_assembly.items());
```
Comment (chriseth) [inline libevmasm/ConstantOptimiser.cpp]: Ah ok, if you also need a const_cast here, then the other version is of course better. Sorry, I though the const_cast would only be needed for the tests.

Comment (axic) [reply libevmasm/ConstantOptimiser.cpp]: Changed again, though this time using `items` and not `itemsMutable`. Let me know which one is preferred.

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
Sorry for the hassle!

**Pull request merged**